gpcontrib/arenadata_toolkit/sql/upgrade_test.sql

```diff
@@ -0,0 +1,145 @@
+-- Cleanup before next step
+	DROP EXTENSION arenadata_toolkit;
+	DROP SCHEMA arenadata_toolkit CASCADE;
+
+-- Create, adb_create_tables and alter extension:
+	CREATE EXTENSION arenadata_toolkit VERSION '1.0';
+	IF from_version != '1.0'
+	THEN
+		EXECUTE FORMAT($fmt$ALTER EXTENSION arenadata_toolkit
+							UPDATE TO %1$I;$fmt$, from_version);
+	END IF;
+	PERFORM arenadata_toolkit.adb_create_tables();
+	ALTER EXTENSION arenadata_toolkit UPDATE;
+
+-- Check the result
+	IF (SELECT default_version = installed_version
+		FROM pg_available_extensions
+		WHERE name='arenadata_toolkit')
+	THEN
+		RETURN NEXT from_version || ': alter and create_tables check';
+	END IF;
+
+-- Cleanup before next step
+	DROP EXTENSION arenadata_toolkit;
+	DROP SCHEMA arenadata_toolkit CASCADE;
+
+-- Create, adb_create_tables, adb_collect_table_stats and alter extension:
+	CREATE EXTENSION arenadata_toolkit VERSION '1.0';
+	IF from_version != '1.0'
+	THEN
+		EXECUTE FORMAT($fmt$ALTER EXTENSION arenadata_toolkit
+						UPDATE TO %1$I;$fmt$, from_version);
+	END IF;
+	PERFORM arenadata_toolkit.adb_create_tables();
+	PERFORM arenadata_toolkit.adb_collect_table_stats();
+	ALTER EXTENSION arenadata_toolkit UPDATE;
+
+-- Check the result
+	IF (SELECT default_version = installed_version
+		FROM pg_available_extensions
+		WHERE name='arenadata_toolkit')
+	THEN
+		RETURN NEXT from_version || ': alter, create_tables and collect_table_stats check';
+	END IF;
+
+-- Check field "tablespace_location" and table "db_files_history_backup_YYYYMMDDtHHMMSS"
+-- which were added at version 1.4
+	PERFORM arenadata_toolkit.adb_create_tables();
+
+	IF 4 = (SELECT count(1)
+			FROM pg_class c
+			LEFT JOIN pg_namespace n ON n.oid = c.relnamespace
+			LEFT JOIN pg_attribute a ON a.attrelid = c.oid AND
+			                            a.attname = 'tablespace_location'
+			WHERE n.nspname = 'arenadata_toolkit' AND
+			      c.relname IN ('db_files_current',
+			                    '__db_files_current',
+			                    '__db_files_current_unmapped',
+			                    'db_files_history') AND
+			      a.attrelid IS NOT NULL)
+	THEN
+		RETURN NEXT from_version || ': column tablespace_location check';
+	END IF;
+
+	IF EXISTS (SELECT 1
+			   FROM pg_tables
+			   WHERE schemaname='arenadata_toolkit' AND
+			         tablename SIMILAR TO 'db_files_history_backup_[0-9]{8}t[0-9]{6}')
+	THEN
+		RETURN NEXT from_version || ': db_files_history_backup check';
+	END IF;
+
+-- Cleanup before next step
+	DROP EXTENSION arenadata_toolkit;
+	DROP SCHEMA arenadata_toolkit CASCADE;
+
+-- Check create extension with the latest version after current was installed and dropped
+	CREATE EXTENSION arenadata_toolkit VERSION '1.0';
+	IF from_version != '1.0'
+	THEN
+		EXECUTE FORMAT($fmt$ALTER EXTENSION arenadata_toolkit
+						UPDATE TO %1$I;$fmt$, from_version);
+	END IF;
+	PERFORM arenadata_toolkit.adb_create_tables();
+	PERFORM arenadata_toolkit.adb_collect_table_stats();
+	DROP EXTENSION arenadata_toolkit;
```

Comment:
Why should I do that? This code checks, that toolkit will be installed correctly, even if there are some relations from previous installation.

Reply:
You removed all from that commit except pointed code

Reply:
This code is not related to code, which removed temporary tables.

Reply:
> What does this code check?

After install and remove toolkit, there are some relations, which was created by toolkit's functions "adb_create_tables" and "adb_collect_table_stats". These relations are not removed automaticly. And this code checks, that next installation will be success.

> And why was it added with said commit?

Maybe, I forgot to add short comment to commit message about this code.

Reply:
> there are some relations, which was created by toolkit's functions "adb_create_tables" and "adb_collect_table_stats"

But aren’t they deleted higher up when the schema is deleted in cascade?

Reply:
> this code checks, that next installation will be success.

Doesn't the code below check this?

Reply:
> And this code checks, that next installation will be success.

red1452 is talking about code below

Reply:
> But aren’t they deleted above up when the schema is deleted in cascade?

This code checks the situation, when command "DROP SCHEMA arenadata_toolkit CASCADE;" is not executed.

Reply:
> This part is preparation for testing:
> 
> ```sql
> 	CREATE EXTENSION arenadata_toolkit VERSION '1.0';
> 	IF from_version != '1.0'
> 	THEN
> 		EXECUTE FORMAT($fmt$ALTER EXTENSION arenadata_toolkit
> 						UPDATE TO %1$I;$fmt$, from_version);
> 	END IF;
> 	PERFORM arenadata_toolkit.adb_create_tables();
> 	PERFORM arenadata_toolkit.adb_collect_table_stats();
> 	DROP EXTENSION arenadata_toolkit;
> ```

Why wasn't this needed before the commit https://github.com/arenadata/gpdb/commit/f90262d7ea86c4f3abae92ee180022699df16237?

Reply:
> This code checks the situation, when command "DROP SCHEMA arenadata_toolkit CASCADE;" is not executed.

OK, I think you should remove lines 106-107 and 109-117, as it was before commit f90262d.

Reply:
> OK, I think you should remove lines 106-107 and 109-117, as it was before commit [f90262d](https://github.com/arenadata/gpdb/commit/f90262d7ea86c4f3abae92ee180022699df16237).

NO, It will change test's logic: now this part of test checks creating the latest version of extension after each versions . If I remove these lines, test will check creating the latest version of extension after the latest version.

Reply:
> Why wasn't this needed before the commit [f90262d](https://github.com/arenadata/gpdb/commit/f90262d7ea86c4f3abae92ee180022699df16237)?

At first redaction of this patch (where was to arguments at function `do_upgrade_test_for_arenadata_toolkit`: `from_version` and `to_version`) `to_version` was different for some executions, but at the last redaction of this patch `to_version` is the last version of toolkit and does not change at the test.

Reply:
I still think you should remove lines 106-107 and 109-117, because they are redundant

Reply:
Let's compare two variants of test:
At line 67, arenadata_toolkit is updated to the latest version. At line 105-106 areanadata_toolkit is dropped and all relations are removed. At lines 110-114 arenadata_toolkit is created with version `from_version` (it is important). All waste will be created at version `from_version` of arenadata_toolkit. If I remove lines 106-107 and 109-117 than for each execution of `do_upgrade_test_for_arenadata_toolkit` check at lines 119-129 will do the same things, because waste will be from the latest version of arenadata_toolkit.

Reply:
> But where are same tests for other cases

I think, It can not check more than my case (old -> create tables -> collect stats -> drop -> create extension), but the test will be done longer without any helpful result. Now, this test is the longest test at arenadata_toolkit.

Reply:
>old -> drop -> create extension
old -> create tables -> drop -> create extension

We definitely should rework the upgrade tests framework in future. I bet we may leave here only tests related to 1.4 specification. Other general cases like above would be introduced when we add an extensible version of upgrade tests. What do you think?

Reply:
Previously, the comment on line 108 referred to the code on lines 109-117, but now this code is left without a comment.